### ISSUE TITLE
about channel order of input images

### ISSUE THREAD

**Issue by FloydEdwin** [facebookresearch/video-long-term-feature-banks#35]
hello! 
I have a question about your caffe2's  input images. Is that BGR or RGB(such as using_BGR=false)?Thank you!

**Comment by chaoyuaw** [comment]
Hi @fountain-K, It's RGB. 

Thanks!

**Comment by FloydEdwin** [comment]
> Hi @fountain-K, It's RGB.
> 
> Thanks!
Thank you for your excellent work！I find in res5 ,the net output size is [16x14x14] ,however in the NL or other 3D_resnet it was 16x7x7.It seems to remain more spatial informations. Can you tell me the reason for your change? Thanks!

**Comment by chaoyuaw** [comment]
Hi @fountain-K , We followed https://eccv2018.org/openaccess/content_ECCV_2018/papers/Xiaolong_Wang_Videos_as_Space-Time_ECCV_2018_paper.pdf
for the architecture, as this paper also pools local information based on bounding boxes. Using more spatial dimensions for tasks that require more spatial details to solve (e.g. for AVA we need to have representations for each person) seems reasonable to me.

**Comment by FloydEdwin** [comment]
> Hi @fountain-K , We followed https://eccv2018.org/openaccess/content_ECCV_2018/papers/Xiaolong_Wang_Videos_as_Space-Time_ECCV_2018_paper.pdf
> for the architecture, as this paper also pools local information based on bounding boxes. Using more spatial dimensions for tasks that require more spatial details to solve (e.g. for AVA we need to have representations for each person) seems reasonable to me.

Thank you for you reply! I asked you there questions because i want to transfer model to pytorch.
I start my work on R50-I3D-NL download from [Getting started.Download pre-trained weights].

I want to know the  R50/101-I3D-NL's accuracy on Kinetics-400. I only get 59% top1 about R50.
I think what confused me most are  the stride , padding values and the model arch.

In LFB paper, It said the model achieves 74.9% (91.6%) top-1 (top-5) accuracy on
the Kinetics-400 .It was the same result on NL paper. So i want to know:
1. Is the **r50_k400_pretrained.pkl** the same as model in NL github（**i3d_nonlocal_32x2_IN_pretrain_400k.pkl**）?
2. the NL output is 7x7 in res5, but it was 14x14 in Table 5 from Appendix A ,so when we finetune the network to other dataset. we change the stride in res5 directly so that we get the 14x14 size output?
3.

> Finally, we remove the temporal striding for conv1 and pool1. 

 I can not undersatnd this sentence  well from the paper .For example, conv1's stride is always the (5,7,7) ,i don't know what 's the means of remove? 

Looking forward to your reply! Thank you!

**Comment by chaoyuaw** [comment]
Hi @fountain-K, 

Sorry for the confusion. Our pre-trained model was trained on Kinetics using the architecture defined in the original non-local paper (with 4x7x7 output) and that model achieved 74.9. Please see https://github.com/facebookresearch/video-nonlocal-net for the exact architecture. We directly take the weight of that 4x7x7-output model trained on Kinetics to initialize our 16x14x14-ouput models. 

I think the most accurate way to see the difference is comparing
https://github.com/facebookresearch/video-nonlocal-net/blob/master/lib/models/resnet_video_org.py
and 
https://github.com/facebookresearch/video-long-term-feature-banks/blob/master/lib/models/model_builder_video.py

and also comparing
Table 1 of https://arxiv.org/pdf/1711.07971.pdf
and 
Appendix A of https://arxiv.org/pdf/1812.05038.pdf

Maybe it also worths checking https://github.com/Tushar-N/pytorch-resnet3d
I did not read too carefully, but my understanding is that Tushar's repository does what you're trying to do for Kinetics, and should work with the pre-trained models we provide here. 

Also just in case you missed it, there is a related discussion in 
https://github.com/facebookresearch/video-long-term-feature-banks/issues/31

As for your original questions: 
1. It's not from the same run, but should have the same architecture as the "run_i3d_nlnet_400k_32f.sh" model in NL repo.
2. stride of res5 is one of the differences. Please see the codes and papers (as mentioned above) for other differences. 
3. (5, 7, 7) is the kernel size, not stride. In NL paper, the output of conv1 is 16×112×112 because it has stride 2 in time to downsample from 32. In our paper, the output of conv1 is 32×112×112 because we don't have the stride. I think once you compare the codes and papers (as mentioned above) it'll be clear :)

Please feel free to let me know if you have any additional questions. Thanks!

**Comment by FloydEdwin** [comment]
> Hi @fountain-K,
> 
> Sorry for the confusion. Our pre-trained model was trained on Kinetics using the architecture defined in the original non-local paper (with 4x7x7 output) and that model achieved 74.9. Please see https://github.com/facebookresearch/video-nonlocal-net for the exact architecture. We directly take the weight of that 4x7x7-output model trained on Kinetics to initialize our 16x14x14-ouput models.
> 
> I think the most accurate way to see the difference is comparing
> https://github.com/facebookresearch/video-nonlocal-net/blob/master/lib/models/resnet_video_org.py
> and
> https://github.com/facebookresearch/video-long-term-feature-banks/blob/master/lib/models/model_builder_video.py
> 
> and also comparing
> Table 1 of https://arxiv.org/pdf/1711.07971.pdf
> and
> Appendix A of https://arxiv.org/pdf/1812.05038.pdf
> 
> Maybe it also worths checking https://github.com/Tushar-N/pytorch-resnet3d
> I did not read too carefully, but my understanding is that Tushar's repository does what you're trying to do for Kinetics, and should work with the pre-trained models we provide here.
> 
> Also just in case you missed it, there is a related discussion in
> #31
> 
> As for your original questions:
> 
> 1. It's not from the same run, but should have the same architecture as the "run_i3d_nlnet_400k_32f.sh" model in NL repo.
> 2. stride of res5 is one of the differences. Please see the codes and papers (as mentioned above) for other differences.
> 3. (5, 7, 7) is the kernel size, not stride. In NL paper, the output of conv1 is 16×112×112 because it has stride 2 in time to downsample from 32. In our paper, the output of conv1 is 32×112×112 because we don't have the stride. I think once you compare the codes and papers (as mentioned above) it'll be clear :)
> 
> Please feel free to let me know if you have any additional questions. Thanks!


Thank you very much. I think I have known what happend!
The [Table5](https://arxiv.org/pdf/1812.05038.pdf) ,we take conv1 as a example.
conv1 5×7×7, 64, stride 1, 2, 2
in fact we do pretrained on kinetics ,we used stride(2,2,2)[not 1,2,2] 
when we to do the finetune work. we change stride to 1(also res5 ),but take the weight of before.(7x7 output)
That is what i confused sentence:

> Finally, we remove the temporal striding for conv1 and pool1.

now i have get the right result of r50.r101 hasn't been verified.
OK, Now I want to know, why not change the networks strides(like table5) to pretrain on knetics400,and then to finetune on other datasets? I fell that may make the **initialization match better**?For me,I don't have so many GPUS to do this work.
So the reason do  strides changes rather than start a new pretrain work is to save time?I know pretrained will cost about at least two weeks.

**Comment by chaoyuaw** [comment]
Hi @fountain-K , I agree. I think pre-training Kinetics using the same architecture might improve performance. The main reason why we didn't do this is because we followed the "Videos as Space-Time Region Graphs" paper, which also didn't do this. I also didn't worry about this too much because our focus is to study the impact of the long-term information, and improving pre-training seems like a close-to-orthogonal direction to me.

**Comment by FloydEdwin** [comment]
> Hi @fountain-K , I agree. I think pre-training Kinetics using the same architecture might improve performance. The main reason why we didn't do this is because we followed the "Videos as Space-Time Region Graphs" paper, which also didn't do this. I also didn't worry about this too much because our focus is to study the impact of the long-term information, and improving pre-training seems like a close-to-orthogonal direction to me.

I agree with you.Spend time on pretrained is not a wise choice.Thank you again for your prompt reply.
